File: server/internal/models/messages.go
package models

import (
	"encoding/json"
	"fmt"
	"sync"
	"time"
)

// MessageType represents the type of WebSocket message
type MessageType string

const (
	// Client to Server message types
	MessageTypeExecute         MessageType = "execute"
	MessageTypeProjectCreate   MessageType = "project_create"
	MessageTypeProjectDelete   MessageType = "project_delete"
	MessageTypeProjectList     MessageType = "project_list"
	MessageTypeProjectJoin     MessageType = "project_join"
	MessageTypeProjectLeave    MessageType = "project_leave"
	MessageTypeAgentNewSession MessageType = "agent_new_session"
	MessageTypeAgentKill       MessageType = "agent_kill"
	MessageTypeGetMessages     MessageType = "get_messages"

	// Server to Client message types
	MessageTypeError            MessageType = "error"
	MessageTypeProjectState     MessageType = "project_state"
	MessageTypeProjectJoined    MessageType = "project_joined"
	MessageTypeProjectLeft      MessageType = "project_left"
	MessageTypeAgentMessage     MessageType = "agent_message"
	MessageTypeServerStats      MessageType = "server_stats"
	MessageTypeHealthCheck      MessageType = "health_check"
	MessageTypeProjectUpdate    MessageType = "project_update"
	MessageTypeProjectDeleted   MessageType = "project_deleted"
	MessageTypeSessionReset     MessageType = "session_reset"
	MessageTypeProcessKilled    MessageType = "process_killed"
	MessageTypeConnectionHealth MessageType = "connection_health"
)

// ClientMessage represents a message from client to server
type ClientMessage struct {
	Type      MessageType     `json:"type"`
	ProjectID string          `json:"project_id,omitempty"`
	Data      json.RawMessage `json:"data,omitempty"`
}

// ServerMessage represents a message from server to client
type ServerMessage struct {
	Type      MessageType `json:"type"`
	ProjectID string      `json:"project_id,omitempty"`
	Data      interface{} `json:"data"`
}

// ExecuteCommand contains parameters for executing Claude
type ExecuteCommand struct {
	Prompt  string         `json:"prompt"`
	Options *ClaudeOptions `json:"options,omitempty"`
}

// ClaudeOptions contains optional parameters for Claude execution
type ClaudeOptions struct {
	DangerouslySkipPermissions bool     `json:"dangerously_skip_permissions,omitempty"`
	AllowedTools               []string `json:"allowed_tools,omitempty"`
	DisallowedTools            []string `json:"disallowed_tools,omitempty"`
	MCPConfig                  string   `json:"mcp_config,omitempty"`
	AppendSystemPrompt         string   `json:"append_system_prompt,omitempty"`
	PermissionMode             string   `json:"permission_mode,omitempty"`
	Model                      string   `json:"model,omitempty"`
	FallbackModel              string   `json:"fallback_model,omitempty"`
	AddDirs                    []string `json:"add_dirs,omitempty"`
	StrictMCPConfig            bool     `json:"strict_mcp_config,omitempty"`
}

// ProjectCreateData contains data for creating a project
type ProjectCreateData struct {
	Path string `json:"path"`
}

// ProjectJoinData contains data for joining a project
type ProjectJoinData struct {
	ProjectID string `json:"project_id"`
}

// GetMessagesData contains parameters for retrieving message history
type GetMessagesData struct {
	Since time.Time `json:"since,omitempty"`
	Limit int       `json:"limit,omitempty"`
}

// ErrorData contains error information sent to clients
type ErrorData struct {
	Code    string                 `json:"code"`
	Message string                 `json:"message"`
	Details map[string]interface{} `json:"details,omitempty"`
}

// ProjectStateData contains project state information
type ProjectStateData struct {
	ID         string    `json:"id"`
	Path       string    `json:"path"`
	State      State     `json:"state"`
	SessionID  string    `json:"session_id,omitempty"`
	CreatedAt  time.Time `json:"created_at"`
	LastActive time.Time `json:"last_active"`
}

// ServerStatsData contains server statistics
type ServerStatsData struct {
	ActiveConnections int       `json:"active_connections"`
	ActiveProjects    int       `json:"active_projects"`
	Uptime            string    `json:"uptime"`
	Timestamp         time.Time `json:"timestamp"`
}

// HealthCheckData contains health check information
type HealthCheckData struct {
	Status      string                 `json:"status"` // "healthy", "degraded", "unhealthy"
	ClaudeReady bool                   `json:"claude_ready"`
	Resources   map[string]interface{} `json:"resources"`
	Timestamp   time.Time              `json:"timestamp"`
}

// TimestampedMessage represents a message with timestamp and direction
type TimestampedMessage struct {
	Timestamp time.Time     `json:"timestamp"`
	Message   ClaudeMessage `json:"message"`
	Direction string        `json:"direction"` // "client" or "claude"
}

// ClaudeMessage represents a message from Claude CLI
type ClaudeMessage struct {
	Type    string          `json:"type"`
	Content json.RawMessage `json:"content,omitempty"`
	Error   string          `json:"error,omitempty"`
}

// MessageLog is now implemented in the storage package
// This remains here for backward compatibility and as part of the Project struct
type MessageLog struct {
	// Deprecated: Use storage.MessageLog instead
	projectID    string
	logDir       string
	currentFile  string
	messageCount int
	fileSize     int64
	mu           sync.Mutex
}

// Validate validates a ClientMessage
func (m *ClientMessage) Validate() error {
	if m.Type == "" {
		return fmt.Errorf("message type cannot be empty")
	}

	// Validate specific message types
	switch m.Type {
	case MessageTypeExecute:
		if m.ProjectID == "" {
			return fmt.Errorf("project_id required for execute")
		}
		var cmd ExecuteCommand
		if err := json.Unmarshal(m.Data, &cmd); err != nil {
			return fmt.Errorf("invalid execute command data: %v", err)
		}
		if cmd.Prompt == "" {
			return fmt.Errorf("prompt cannot be empty")
		}

	case MessageTypeProjectCreate:
		var data ProjectCreateData
		if err := json.Unmarshal(m.Data, &data); err != nil {
			return fmt.Errorf("invalid project create data: %v", err)
		}
		if data.Path == "" {
			return fmt.Errorf("path cannot be empty")
		}

	case MessageTypeProjectDelete, MessageTypeAgentNewSession, MessageTypeAgentKill:
		if m.ProjectID == "" {
			return fmt.Errorf("project_id required for %s", m.Type)
		}

	case MessageTypeProjectJoin:
		var data ProjectJoinData
		if err := json.Unmarshal(m.Data, &data); err != nil {
			return fmt.Errorf("invalid project join data: %v", err)
		}
		if data.ProjectID == "" {
			return fmt.Errorf("project_id cannot be empty")
		}

	case MessageTypeProjectLeave:
		if m.ProjectID == "" {
			return fmt.Errorf("project_id required for project_leave")
		}

	case MessageTypeGetMessages:
		if m.ProjectID == "" {
			return fmt.Errorf("project_id required for get_messages")
		}
	}

	return nil
}

// NewErrorMessage creates an error message for the client
func NewErrorMessage(projectID, code, message string, details map[string]interface{}) ServerMessage {
	return ServerMessage{
		Type:      MessageTypeError,
		ProjectID: projectID,
		Data: ErrorData{
			Code:    code,
			Message: message,
			Details: details,
		},
	}
}

// NewProjectStateMessage creates a project state update message
func NewProjectStateMessage(project *Project) ServerMessage {
	return ServerMessage{
		Type:      MessageTypeProjectState,
		ProjectID: project.ID,
		Data: ProjectStateData{
			ID:         project.ID,
			Path:       project.Path,
			State:      project.State,
			SessionID:  project.SessionID,
			CreatedAt:  project.CreatedAt,
			LastActive: project.LastActive,
		},
	}
}

// NewServerStatsMessage creates a server statistics message
func NewServerStatsMessage(connections, projects int, uptime time.Duration) ServerMessage {
	return ServerMessage{
		Type: MessageTypeServerStats,
		Data: ServerStatsData{
			ActiveConnections: connections,
			ActiveProjects:    projects,
			Uptime:            uptime.String(),
			Timestamp:         time.Now(),
		},
	}
}

File: frontend-spa/src/types/messages.ts
// Configuration types
export interface MCPConfig {
  [key: string]: unknown;
}

// Connection types
export type ConnectionStatus = 'connected' | 'disconnected' | 'connecting' | 'error';
export type ConnectionState = ConnectionStatus; // Alias for compatibility

// Client to Server Messages
export interface ClientMessage {
  type: string;
  project_id?: string;
  data?: Record<string, unknown>;
}

export interface ProjectCreateMessage extends ClientMessage {
  type: 'project_create';
  data: {
    path: string;
  };
}

export interface ExecuteMessage extends ClientMessage {
  type: 'execute';
  project_id: string;
  data: {
    prompt: string;
    options?: {
      model?: string;
      dangerously_skip_permissions?: boolean;
      allowed_tools?: string[];
      disallowed_tools?: string[];
      permission_mode?: string;
      mcp_config?: MCPConfig;
      append_system_prompt?: string;
      fallback_model?: string;
      add_dirs?: string[];
      strict_mcp_config?: boolean;
    };
  };
}

export interface ProjectJoinMessage extends ClientMessage {
  type: 'project_join';
  data: {
    project_id: string;
  };
}

export interface ProjectLeaveMessage extends ClientMessage {
  type: 'project_leave';
  project_id: string;
}

export interface ProjectDeleteMessage extends ClientMessage {
  type: 'project_delete';
  project_id: string;
}

export interface ProjectListMessage extends ClientMessage {
  type: 'project_list';
}

export interface GetMessagesMessage extends ClientMessage {
  type: 'get_messages';
  project_id: string;
  data: {
    since?: string;
    limit?: number;
  };
}

export interface AgentKillMessage extends ClientMessage {
  type: 'agent_kill';
  project_id: string;
}

export interface AgentNewSessionMessage extends ClientMessage {
  type: 'agent_new_session';
  project_id: string;
}

// Server to Client Messages
export interface ServerMessage {
  type: string;
  project_id?: string;
  data: Record<string, unknown>;
  timestamp?: string;
}

export interface ProjectStateMessage extends ServerMessage {
  type: 'project_state';
  project_id: string;
  data: {
    id: string;
    path: string;
    state: 'IDLE' | 'EXECUTING' | 'ERROR';
    session_id?: string | null;
    created_at: string;
    last_active: string;
  };
}

export interface ProjectListResponseMessage extends ServerMessage {
  type: 'project_list_response';
  data: {
    projects: Array<{
      id: string;
      path: string;
      state: string;
      created_at: string;
      last_active: string;
      session_id?: string;
    }>;
  };
}

export interface ProjectDeletedMessage extends ServerMessage {
  type: 'project_deleted';
  project_id: string;
}

export interface ProjectJoinedMessage extends ServerMessage {
  type: 'project_joined';
  project_id: string;
}

export interface ProjectLeftMessage extends ServerMessage {
  type: 'project_left';
  project_id: string;
}

export interface AgentMessage extends ServerMessage {
  type: 'agent_message';
  project_id: string;
  data: {
    type: string;
    [key: string]: unknown;
  };
}

export interface ErrorMessage extends ServerMessage {
  type: 'error';
  data: {
    code: string;
    message: string;
    details?: Record<string, unknown>;
  };
}

export interface SessionResetMessage extends ServerMessage {
  type: 'session_reset';
  project_id: string;
}

export interface MessagesResponseMessage extends ServerMessage {
  type: 'messages_response';
  project_id: string;
  data: {
    messages: Array<{
      timestamp: string;
      direction: 'client' | 'claude';
      message: string | Record<string, unknown>;
    }>;
  };
}

export interface HealthCheckMessage extends ServerMessage {
  type: 'health_check';
  data: {
    status: 'healthy' | 'degraded' | 'unhealthy';
    uptime: number;
    version: string;
    connections: {
      active: number;
      limit: number;
    };
    projects: {
      count: number;
      limit: number;
      executing: number;
    };
    resources: {
      cpu_percent: number;
      memory_mb: number;
      goroutines: number;
      disk_free_gb: number;
    };
    claude: {
      available: boolean;
      version: string;
    };
  };
}

export interface ServerStatsMessage extends ServerMessage {
  type: 'server_stats';
  data: {
    connections: number;
    projects: number;
    executions: {
      active: number;
      total: number;
      success: number;
      failed: number;
      avg_duration_ms: number;
    };
    messages: {
      sent: number;
      received: number;
      rate_per_sec: number;
    };
    uptime: number;
  };
}

// Enhanced project creation messages for Task 8: WebSocket project creation integration
// Note: Using the correct ProjectCreateMessage interface above

export interface ProjectCreatedMessage extends ServerMessage {
  type: 'project_created';
  data: {
    project: {
      id: string;
      name: string;
      path: string;
      created_at: string;
      state: string;
      session_id?: string;
    };
    server_id: string;
    created_at: string;
  };
}

export interface ProjectCreationErrorMessage extends ServerMessage {
  type: 'project_creation_error';
  data: {
    error_code: string;
    message: string;
    details?: Record<string, unknown>;
    server_id: string;
  };
}

// Additional server message types based on server specification
export interface ProjectUpdateMessage extends ServerMessage {
  type: 'project_update';
  project_id: string;
  data: {
    update_type: string;
    details?: Record<string, unknown>;
  };
}

export interface ProcessKilledMessage extends ServerMessage {
  type: 'process_killed';
  project_id: string;
  data: {
    process_id?: string;
    reason?: string;
    exit_code?: number;
  };
}

export interface ConnectionHealthMessage extends ServerMessage {
  type: 'connection_health';
  data: {
    status: 'healthy' | 'degraded' | 'unhealthy';
    latency_ms?: number;
    last_heartbeat?: string;
    connection_quality?: number;
  };
}

// Error codes
export enum ErrorCode {
  INVALID_MESSAGE = 'INVALID_MESSAGE',
  INVALID_PATH = 'INVALID_PATH',
  PROJECT_NESTING = 'PROJECT_NESTING',
  PROJECT_NOT_FOUND = 'PROJECT_NOT_FOUND',
  PROJECT_LIMIT = 'PROJECT_LIMIT',
  EXECUTION_TIMEOUT = 'EXECUTION_TIMEOUT',
  CLAUDE_NOT_FOUND = 'CLAUDE_NOT_FOUND',
  PROCESS_ACTIVE = 'PROCESS_ACTIVE',
  RESOURCE_LIMIT = 'RESOURCE_LIMIT',
  INTERNAL_ERROR = 'INTERNAL_ERROR',
  // Project creation specific error codes
  PROJECT_CREATION_FAILED = 'PROJECT_CREATION_FAILED',
  PROJECT_VALIDATION_ERROR = 'PROJECT_VALIDATION_ERROR',
  PROJECT_PATH_EXISTS = 'PROJECT_PATH_EXISTS',
  PROJECT_NAME_TAKEN = 'PROJECT_NAME_TAKEN',
  SERVER_UNAVAILABLE = 'SERVER_UNAVAILABLE',
  CONNECTION_TIMEOUT = 'CONNECTION_TIMEOUT',
  PERMISSION_DENIED = 'PERMISSION_DENIED',
}
